packages/editor/src/components/inserter/menu.js

```diff
@@ -369,24 +368,68 @@ export default compose(
 			hideInsertionPoint,
 		} = dispatch( 'core/editor' );
 
+		// To avoid duplication, getInsertionPoint is extracted and used in two event handlers
+		// This breaks the withDispatch not containing any logic rule.
+		// Since it's a function only called when the event handlers are called,
+		// it's fine to extract it.
+		// eslint-disable-next-line no-restricted-syntax
+		function getInsertionPoint() {
+			const {
+				getBlockIndex,
+				getBlockRootClientId,
+				getBlockSelectionEnd,
+				getBlockOrder,
+			} = select( 'core/editor' );
+			const { clientId, rootClientId, isAppender } = ownProps;
+
+			// If the clientId is defined, we insert at the position of the block.
+			if ( clientId ) {
+				return {
+					index: getBlockIndex( clientId, rootClientId ),
+					rootClientId,
+				};
+			}
+
+			// If there a selected block, we insert after the selected block.
+			const end = getBlockSelectionEnd();
```

Comment:
`getBlockSelectionEnd` is performant enough, but there are optimization opportunities here to avoid calling it when `isAppender` is true (allow the [short-circuiting](https://en.wikipedia.org/wiki/Short-circuit_evaluation) of the following condition's `! isAppender`).

```js
let end;
if ( ! isAppender && ( end = getBlockSelectionEnd() ) ) {
	const selectedBlockRootClientId = getBlockRootClientId( end ) || undefined;
	return {
		index: getBlockIndex( end, selectedBlockRootClientId ) + 1,
		rootClientId: selectedBlockRootClientId,
	};
}
```

Or:

```js
if ( ! isAppender ) {
	const end = getBlockSelectionEnd();
	if ( end ) {
		const selectedBlockRootClientId = getBlockRootClientId( end ) || undefined;
		return {
			index: getBlockIndex( end, selectedBlockRootClientId ) + 1,
			rootClientId: selectedBlockRootClientId,
		};
	}
}
```

Reply:
Oh, I forgot this is only called on specific user interactions, not in a `withSelect`, so not nearly as big a deal.